src/main/java/org/purescript/file/ExportedTypesIndex.kt

```diff
@@ -0,0 +1,76 @@
+package org.purescript.file
+
+import com.intellij.openapi.application.ReadAction
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiManager
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.util.indexing.*
+import com.intellij.util.io.EnumeratorStringDescriptor
+import org.purescript.psi.PSForeignDataDeclaration
+import org.purescript.psi.data.PSDataDeclaration
+import org.purescript.psi.exports.PSExportedData
+import org.purescript.psi.newtype.PSNewTypeDeclarationImpl
+import org.purescript.psi.typesynonym.PSTypeSynonymDeclaration
+
+/**
+ * An index on what type declarations every module exports.
+ * The index contains the following declarations:
+ *  - [PSTypeSynonymDeclaration]
+ *  - [PSDataDeclaration]
+ *  - [PSNewTypeDeclarationImpl]
+ *  - [PSForeignDataDeclaration]
+ */
+class ExportedTypesIndex : ScalarIndexExtension<String>(), DataIndexer<String, Void?, FileContent> {
+
+    override fun map(inputData: FileContent): Map<String, Void?> {
+        val file = inputData.psiFile as? PSFile ?: return emptyMap()
+        val module = file.module ?: return emptyMap()
+        val exportList = module.exportList
+        return if (exportList == null) {
+            val exportedNames = mutableSetOf<String>()
+            module.typeSynonymDeclarations.mapTo(exportedNames) { it.name }
+            module.dataDeclarations.mapTo(exportedNames) { it.name }
+            module.newTypeDeclarations.mapTo(exportedNames) { it.name }
+            module.foreignDataDeclarations.mapTo(exportedNames) { it.name }
+            exportedNames.associateWith { null }
+        } else {
+            exportList.exportedItems
+                .filterIsInstance<PSExportedData>()
+                .associate { it.name to null }
+        }
+    }
```

Comment:
I think the code becomes cleaner if we implement the `DataIndexer` interface directly. That way we don't have the large anonymous function in the index.